src/goto-instrument/contracts/contracts.cpp

```diff
@@ -890,6 +890,20 @@ void code_contractst::apply_loop_contract(
       throw 0;
     }
 
+    // Add a skip before loop head and let all goto's (except for the loop end)
+    // that jumps to loop head target the skip instead.
+    const auto it_before_loop_head =
+      goto_function.body.insert_before(loop_head, goto_programt::make_skip());
+    for(auto it = goto_function.body.instructions.begin();
```

Comment:
Actually, we cannot simply use jump-preserving insertion to insert instructions before ```loop_head``` in the function ```check_apply_loop_contracts```. We use ```loop``` to store all instructions in the original loop so that we can correctly instrument assign checks for only original instructions (https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/contracts/contracts.cpp#L397), or we can decide whether a original jump jumping out of the loop (https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/contracts/contracts.cpp#L474).

However, after ```insert_before_swap_and_advance(goto_function.body, loop_end, new_instrs)````, the ```loop_head``` will not be in ```loop``` anymore. It will cause problems such as not instrumenting assignable checks for variables in loop guards, or not correctly check the loop instrumentation was not truncated.

So it is not safe to use jump-preserving insertions on original instructions in the loops.